mantidproject/mantid
# Create standalone Windows installer using conda environment

**Description of work.**
Create a standalone Windows installer from a conda package. This includes:
- a bash script that creates a new conda environment and installs mantid workbench from whichever chanel you specify (e.g. pass the argument `-c mantid/label/nightly` to get the nightly package)
- an NSIS script that replicates the behaviour of our old Windows packaging solution, but instead essentially packages up the files from the conda environment
- a python helper script to ensure that we only uninstall files that are installed



### **To test:**

**Part 1 - generate the installer**
1. Open your miniforge prompt.
2. Ensure you have an up-to-date mambaforge installation (e.g. with `mamba update mamba`), otherwise the following may not work.
3. Check that NSIS isn't installed on your system (either try running `nsis` in a command prompt, or check the Windows "Apps & Features" list) - otherwise it won't be a fair test.
4. Run `create_package.sh` using the following, but replacing `C:/Users/sam/mantid/installers/conda/win/create_package.sh` with your path to the script in the mantid repo:
`"C:\Program Files\Git\bin\bash.exe" -c "C:/Users/sam/mantid/installers/conda/win/create_package.sh -c mantid/label/nightly -s Unstable MantidWorkbenchUnstable.exe"`
this sould make an exe called `MantidWorkbenchUnstable.exe` with the unstable icon. It will use the latest uploaded nightly from the conda channels.
5. You could also try running the above with `-s Nightly` instead of -s Unstable, which should generate an installer with the nightly icons.
6. Verify that the only extra files left behind are the installer and the nsis log. All other temporary files should be removed by the script.

**Part 2 - test the installer/uninstaller**
1. Run the installer generated in **part 1**, installing to a location of your choosing.
2. Try opening all of the following (making sure that you use the `unstable` labelled version if that is what you generated):
    - MantidWorkbench from the desktop and start menu shortcuts
    - Mantid notebook from the desktop and start menu
3. With Mantid open, select a default facility and instrument, and change some settings via the menu in File->Settings. Make a note of these non-default settings.
4. Check the icon in Windows `Apps & Features` (the Add/Remove programs list), it should be a mantid with the appropriate label (e.g unstable)
5. Run the uninstaller
    - Check that the desktop and start menu shortcuts are removed, and that it's not longer listed in Windows `Apps & Features`.
    - Check that the directory you installed Mantid to is empty except for the `Mantid.user.properties` file.
7. Run the installer again and open Mantid to check that the settings that you chose in step 3 have persisted (i.e. the `Mantid.user.properties` file wasn't removed by uninstalling)
8. With Mantid open, attempt to run the installer. Click through and try to install Mantid into the same directory as before.
    - **It should complain that it cannot continue because Mantid is already running.**
9. Close Mantid and try again. Verify that the installer runs the uninstaller followed by installing.
    - It should perform the extra uninstall step without any extra user intervention.

**Part 3 - test the error reporter**
The error reporter fix won't have been propagated to the nightly channel since it's a part of this PR so to test it you will need to run mantid from your source:
1. Find the Mantid.properties folder inside your build/bin/<config>/ directory (e.g. <config> is RelWithDebInfo)
2. Set usagereports.enabled = 1
3. Run mantid workbench from you source (i.e. using Pycharm or Visual Studio).
4. Run the Segfalut alogithm from the algorithm widget, making sure to untick the dry run option.
5. Wait a few seconds and verify that the error reporter appears.

Fixes #33522
Fixes #33634


*This does not require release notes* because **this change should be invisible to the end user on Windows**




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
